### PR TITLE
fix: render yearn tx transfer type

### PR DESCRIPTION
**Pull request by GMSteuart**
## Description

Updated and added functionality to more properly render Yearn TX transfer types and more generically, any TX that follows vault method naming conventions.

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

fixes issue #1174 

## Risk

The main concern would be with the ever growing complexity of the useTxDetails function and associated objects.

## Testing

I've added two additional yearn transactions for mocking purposes but have omitted the file I used to utilize the mocked yearn transactions since it is out of the scope for this PR and would introduce new development processes.

### UNI
Screen Shot 2022-03-20 at 6 35 12 AM

### yvUNI
Screen Shot 2022-03-20 at 6 36 31 AM

### CODE REVIEW

File: src/components/Transactions/TransactionContract.tsx
Comment (0xApotheosis) [inline src/components/Transactions/TransactionContract.tsx:76]: I'd prefer if we didn't load up `useTxDetails` with `negated` - `TransactionContract` should have all of the info it needs to do the equivalent logic of `(tx.data.method === 'withdraw' && activeTx.type === 'send')`, and then modify the `prefix` accordingly.

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:161]: We shouldn't need to do this.

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:75]: Why is this necessary?

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:86]: We can also probably make this clearer with something like:

```ts
const isReceive = activeTx.type === TxType.Receive
i18n = isReceive ? TxType.Receive : tx.data.method ?? 'unknown'
```

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:77]: As `i18n` is only used in `TransactionContract`, and will likely only ever be, can we move this logic there?

File: src/test/mocks/txs.ts
Comment (0xApotheosis) [inline src/test/mocks/txs.ts:345]: This const breaks lint.

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:84]: Rather than documenting with comments, can we instead prove this is the case using tests?

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Looking good @GMSteuart, I've left some comments to consider.

Also, I notice a `createMockEthTxs` factory was added, but never used - are there tests you intended to commit?

**Comment by GMSteuart** [comment]
@0xApotheosis `createMockEthTxs` can now be seen in a couple of tests for some helper functions found in `useTxDetails`. I had also used it to mock part of my transactions so I can see txs locally without needing to make them on a higher environment network.

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.ts:59]: We should still check for `undefined` here, else we might cast `undefined` as `ContractMethods`.

File: src/components/TransactionHistoryRows/TransactionContract.tsx
Comment (0xApotheosis) [inline src/components/TransactionHistoryRows/TransactionContract.tsx:30]: This is a little strange - the second item of this array is always `false`, which is a hint that we should refactor.

We can change this to something like `firstAssetOutgoing`, which is of type `boolean`. Then in `TransactionGenericRow` we can check if we are mapping over the first asset, and if so, add a `-` to the `prefix`.

File: src/components/Transactions/TransactionContract.tsx
Comment (0xApotheosis) [inline src/components/Transactions/TransactionContract.tsx:36]: The naming here is a little off - this is more like `interactsWithWithdrawMethod`.

File: src/components/Transactions/TransactionContract.tsx
Comment (0xApotheosis) [inline src/components/Transactions/TransactionContract.tsx:39]: `negated` isn't quite right - be more descriptive: `sendInteractsWithWithdrawMethod` or something.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Thanks for the updates @GMSteuart! 

I have a few more requests to tidy things up a little.

File: src/hooks/useTxDetails/useTxDetails.test.ts
Comment (GMSteuart) [reply src/hooks/useTxDetails/useTxDetails.test.ts]: Removed. I had added the type because I was reading documentation somewhere that passing the type ensured the compared values were also the same type.

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (GMSteuart) [reply src/hooks/useTxDetails/useTxDetails.ts]: The plural to singular change is an easy one to forget

File: src/hooks/useTxDetails/useTxDetails.ts
Comment (GMSteuart) [reply src/hooks/useTxDetails/useTxDetails.ts]: Added a test for this. Casting `undefined` as `ContractMethod` won't add it to the type and resulting list.

File: src/components/TransactionHistoryRows/TransactionContract.tsx
Comment (GMSteuart) [reply src/components/TransactionHistoryRows/TransactionContract.tsx]: This is less memory intensive! I had gone the least intrusive route.

File: src/hooks/useTxDetails/useTxDetails.test.ts
Comment (0xApotheosis) [inline src/hooks/useTxDetails/useTxDetails.test.ts:98]: `.toBe(false)`

File: src/hooks/useTxDetails/useTxDetails.test.ts
Comment (0xApotheosis) [reply src/hooks/useTxDetails/useTxDetails.test.ts:93]: This is still a falsy check, please fix.

File: src/hooks/useTxDetails/useTxDetails.test.ts
Comment (0xApotheosis) [reply src/hooks/useTxDetails/useTxDetails.test.ts:84]: This is still a truthy check, please fix.

File: src/components/TransactionHistoryRows/TransactionContract.tsx
Comment (0xApotheosis) [inline src/components/TransactionHistoryRows/TransactionContract.tsx:30]: My naming suggestion may have led you astray here - it should probably be `isAssetOutgoing`, as it's not always the first asset, yeah?

Indeed, you check if it is first in the generic row (`prefix={index === 0 && isFirstAssetOutgoing ? '-' : ''}`).

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]

**Comment by 0xApotheosis** [comment]
We'll also need to resolve rendering in the compact view, as it looks a little rough (though expanded is nice):

Screen Shot 2022-03-23 at 9 37 32 am

I suggest reaching out to product in the public product channel on Discord for feedback/direction.

**Comment by 0xApotheosis** [comment]
> We'll also need to resolve rendering in the compact view, as it looks a little rough (though expanded is nice):

Never mind, this is already kind of broken in `develop` and is not an issue with this PR - ignore.

File: src/components/TransactionHistoryRows/TransactionContract.tsx
Comment (GMSteuart) [reply src/components/TransactionHistoryRows/TransactionContract.tsx:30]: It will always be the first asset in the components assets list (and whats displayed) but not guaranteed to be the first asset in the transfers list where they come from.

**Review by 0xApotheosis — APPROVED** [review]
Nice one @GMSteuart, meets the AC 🥳 

Over to you @0xdef1cafe for a +1.

**Comment by 0xApotheosis** [comment]
@GMSteuart you'll need to fix lint - `yarn lint:fix`.

**Comment by GMSteuart** [comment]
@0xApotheosis linted. Anything preventing adding linting as a precommit hook?

**Comment by 0xApotheosis** [comment]
> @0xApotheosis linted. Anything preventing adding linting as a precommit hook?

A potentially degraded developer experience when committing.

Best to leave it to the developer to whatever suits their workflow - e.g. you can have a watcher running that'll print any lint issues to your terminal in real-time. You could even have it fix issues.

**Comment by 0xApotheosis** [comment]
+ I generally prefer fail-open approaches, allowing constraints to be imposed by the engineer, rather than the other way around.

**Review by 0xApotheosis — APPROVED** [review]

**Comment by 0xdef1cafe** [comment]
hey @GMSteuart, i've got a USDC deposit into a yearn vault that's showing as a generic transaction

image

associated tx ids for reference

https://etherscan.io/tx/0xded9a55622504979d7980b401d3b5fab234c0b64ee779f076df2023929b0f083
https://etherscan.io/tx/0xc9bdf77c7c82f28c34af2ef98b38e75cf201d44af69f245a8e43c497b570620e

you might want to consider rebasing off develop as i believe another transaction related PR went in today

**Comment by GMSteuart** [comment]
@0xdef1cafe I'll take a look at this transaction to see whats going on. I knew I'd need to update the branch at some point since my other pr also touches some of the same transaction files.

**Comment by 0xApotheosis** [comment]
You have a rebase/merge issue ser.

At first glance, I'd wager your fork of `web` is out of date - update it, then merge/rebase again and these issues should dissapear.

**Comment by GMSteuart** [comment]
@0xdef1cafe the chain adapter is failing to parse your tx properly and return a result with the `data` key that would allow the UI to parse it:

```ts
{
    address: '0x934be745172066EDF795ffc5EA9F28f19b440c63',
    blockHash: '0x17d278ffcb1fb940d69e72287339607445d373d0c6a654a61526b0bc805cf10c',
    blockHeight: 13730189,
    blockTime: 1638487560,
    caip2: 'eip155:1',
    chain: ChainTypes.Ethereum,
    confirmations: 729251,
    fee: { caip19: 'eip155:1/slip44:60', value: '9099683709794574' },
    status: chainAdapters.TxStatus.Confirmed,
    tradeDetails: undefined,
    transfers: [
      {
        caip19: 'eip155:1/erc20:0x5f18c75abdae578b483e5f43f12a39cf75b973a9',
        from: '0x0000000000000000000000000000000000000000',
        to: '0x934be745172066EDF795ffc5EA9F28f19b440c63',
        type: chainAdapters.TxType.Receive,
        value: '9178352'
      },
      {
        caip19: 'eip155:1/erc20:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
        from: '0x934be745172066EDF795ffc5EA9F28f19b440c63',
        to: '0x5f18C75AbDAe578b483E5F43f12a39cF75b973a9',
        type: chainAdapters.TxType.Send,
        value: '10000000'
      }
    ],
    txid: '0xded9a55622504979d7980b401d3b5fab234c0b64ee779f076df2023929b0f083'
    // MISSING: data: { method: 'deposit', parser: 'yearn' }
  }
```

**Comment by 0xApotheosis** [comment]
Ah I see - @0xdef1cafe's transaction used the Yearn contract, not the ShapeShift router contract.

We don't currently support the rendering of deposits that interact directly with the Yearn contract. I just whipped up a PR which fixes that: https://github.com/shapeshift/unchained/pull/363

This will render correctly once that PR is merged and deployed.

**Comment by 0xApotheosis** [comment]
The fix noted above is merged, and I have a release branch for it ready to go: https://github.com/shapeshift/unchained/pull/371/

Once merged and deployed @GMSteuart you'll need to update the dependencies in this PR to `6.7.0`.

**Comment by GMSteuart** [comment]
@0xApotheosis I have some other touch ups to do but wanted to at least ask if everything looks good from the package version upgrade?

**Comment by 0xApotheosis** [comment]
> @0xApotheosis I have some other touch ups to do but wanted to at least ask if everything looks good from the package version upgrade?

It looks reasonable, though I didn't expect to see changes to `swagger-ui-dist` in https://github.com/shapeshift/web/pull/1217/commits/cc0c608ea2247efe2aabfa6a588efc17f9d4528b.

Did the new `unchained-client` resolve the issue we found?

**Comment by GMSteuart** [comment]
@0xApotheosis yes, the issue from earlier is solved and appears as a trade and not a transaction. I'll revert those swagger-ui updates.

**Comment by 0xApotheosis** [comment]
@0xdef1cafe a quick re-test with your wallet and the USDC deposit into a Yearn Vault TX and this should be good to merge.

**Comment by 0xdef1cafe** [comment]
image

LGTM!

**Comment by 0xdef1cafe** [comment]
@GMSteuart i fixed the package.json conflict - you might want to pull and run yarn again and commit and `yarn.lock` changes

**Review by 0xdef1cafe — APPROVED** [review]

**Comment by GMSteuart** [comment]
@0xdef1cafe I pulled and ran yarn and there were no changes

**Pull request merged**
